components/script/dom/permissions.rs

```diff
@@ -0,0 +1,237 @@
+    let jsval = ObjectValue(permission_descriptor_obj);
+    property.handle_mut().set(jsval);
+    unsafe {
+        match BluetoothPermissionDescriptor::from_jsval(cx, property.handle(), ()) {
+            Ok(ConversionResult::Success(descriptor)) => Ok(descriptor),
+            Ok(ConversionResult::Failure(message)) => Err(String::from(message)),
+            Err(_) => Err(String::from(BT_DESC_CONVERSION_ERROR)),
+        }
+    }
+}
+
+// https://w3c.github.io/permissions/#dom-permissions-query
+fn sync_default_permission_query_call(global: &GlobalScope,
+                                      descriptor: PermissionDescriptor,
+                                      promise: &Rc<Promise>,
+                                      cx: *mut JSContext) {
+    // Step 5.
+    let status = PermissionStatus::create_from_descriptor(global, PermissionDescriptorType::Default(descriptor));
+    // Step 6.
+    // NOTE: `status.query` is the same as descriptor
+    status.permission_query(&*status.get_query().borrow());
+    // Step 7.
+    promise.resolve_native(cx, &status);
+}
+
+// https://w3c.github.io/permissions/#dom-permissions-request
+fn sync_default_permission_request_call(global: &GlobalScope,
+                                        descriptor: PermissionDescriptor,
+                                        promise: &Rc<Promise>,
+                                        cx: *mut JSContext) {
+    // Step 5.
+    let status = PermissionStatus::create_from_descriptor(global, PermissionDescriptorType::Default(descriptor));
+    // Step 6.
+    // NOTE: `status.query` is the same as descriptor
+    if let Err(err) = status.permission_request(&*status.get_query().borrow()) {
+        // Step 7.
+        promise.reject_error(cx, err);
+    }
+    // Step 8.
+    promise.resolve_native(cx, &status);
+}
+
+// https://w3c.github.io/permissions/#dom-permissions-query
+fn bluetooth_permission_query_call(global: &GlobalScope,
+                                   descriptor: BluetoothPermissionDescriptor,
+                                   promise: &Rc<Promise>,
+                                   cx: *mut JSContext) {
+    // Step 5.
+    let result = BluetoothPermissionResult::create_from_descriptor(
+                    global, PermissionDescriptorType::Bluetooth(descriptor));
+
+    // Step 6.
+    result.permission_query(&promise, cx);
+}
+
+// https://w3c.github.io/permissions/#dom-permissions-request
+fn bluetooth_permission_request_call(global: &GlobalScope,
+                                     descriptor: BluetoothPermissionDescriptor,
+                                     promise: &Rc<Promise>,
+                                     cx: *mut JSContext) {
+    // Step 5.
+    let result =
+        BluetoothPermissionResult::create_from_descriptor(global, PermissionDescriptorType::Bluetooth(descriptor));
+    // Step 6.
+    if let Err(err) = result.permission_request(&promise) {
+        // Step 7.
+        promise.reject_error(cx, err);
+    }
```

Comment:
The resolve for this promise happens in the `handle_response` function [here](https://github.com/szeged/servo/pull/92/files/f82ac8d9620a4a929b79bef596904cbf41bed920#diff-f978d59fbb580e780099d649ab652468R188) or [here](https://github.com/szeged/servo/pull/92/files/f82ac8d9620a4a929b79bef596904cbf41bed920#diff-f978d59fbb580e780099d649ab652468R197) depending on the cached devices. And there is also a return value with this promise at the en of the `Request` call [here](https://github.com/szeged/servo/pull/92/files/f82ac8d9620a4a929b79bef596904cbf41bed920#diff-cb605a27a38dd46b3be20072919a178aR206).